plugins/Core-Extensions/Recommended/edu.toronto.cs.se.modelepedia.operator/src/edu/toronto/cs/se/modelepedia/operator/slice/Slice.java

```diff
@@ -74,20 +76,181 @@ public Input(Map<String, Model> inputsByName) {
 		}
 	}
 
-	protected ModelRel slice(ModelRel critRel, Model model, MID outputMID) throws MMINTException {
 
+	// Extract all model elements from the input slicing criterion. 
+	public List<EObject> getCriterionElements(ModelRel criterion) 
+			throws MMINTException {
+		List<EObject> criterionList = new ArrayList<EObject>();
+ 
+		List<ModelElementReference> merList = 
+				criterion.getModelEndpointRefs().get(0).getModelElemRefs();
+
+		EObject elem;
+		for (ModelElementReference mer : merList) {
+			elem = mer.getObject().getEMFInstanceObject();
+			criterionList.add(elem);
+			System.out.println("Getting input elements: " + elem);
+		}
+		return criterionList;
+	}
+	
+	// Extract all model elements from the model to be sliced.
+	public List<EObject> getAllModelElements(Model in, EObject root) 
+			throws MMINTException {
+		List<EObject> unchecked = new ArrayList<EObject>();
+				
+		Iterator<EObject> iter = root.eAllContents();
+		while (iter.hasNext()) {
+			unchecked.add(iter.next());
+		}
+		
+		return unchecked;
+	}
+
+	// Iterate through the unchecked list of model elements to identify 
+	// all those which are included in the input criterion.
+	// Note: It is assumed that there may be a many-to-many correspondence
+	// between elements in the unchecked list and the criterion list.
+	public List<EObject> matchCriterionToModelElements(
+			List<EObject> unchecked, List<EObject> criterionList) {
+		List<EObject> unprocessed = new ArrayList<EObject>();
+
+		Iterator<EObject> uncheckedIter = unchecked.iterator();
+		while (uncheckedIter.hasNext()) {
+			EObject elem = uncheckedIter.next();
+			for (EObject critElem : criterionList) {
+				if (isModelElemEqual(elem, critElem)) {
+					unprocessed.add(elem);
+					System.out.println("Criterion Added: " + elem);
+					uncheckedIter.remove();
+				}
+			}
+		}
+		
+		return unprocessed;
+	}
+	
+	// Checks whether the two input model elements are equivalent. 
+	// This is necessary as even for the same model elements, different 
+	// calls to getEMFInstanceObject() will return different EMF instances.
+	public boolean isModelElemEqual(EObject obj1, EObject obj2) {
```

Comment:
You are right that getEMFInstanceObject() will return different EMF objects, because each time it'll load a new model in memory. The problem here is that you are trying to compare EMF objects of two models in memory that are really the same model by construction, it's not that they are given and you have to figure it out.
isModelElemEqual should not be needed because there should not be two instances of the same model loaded in memory.